New serverless pattern - s3-lambda-comprehend-dynamodb-batch-sentiment-analysis (#1985)
aws-samples/serverless-patterns: s3-lambda-comprehend-cdk-python/README.md

# S3 -> Lambda -> Comprehend
This patterns shows CDK deployment on how to leverage Amazon S3, AWS Lambda, Amazon Comprehend and Amazon DynamoDB to perform Batch Sentiment Analysis in a serverless fashion.

## Architecture
![Diagram](src/architecture.jpg)

### What resources will be created?
This CDK code will create the following:
   - One S3 bucket (to hold the text for which sentiment analysis has to be done)

>    - One Lambda (to invoke the Comprehend API)

Reviewer: ```suggestion
   - One Lambda function (to invoke the Comprehend API)
```

   - 1 DynamoDB (to store the result of Sentiment Analysis)

>    - 1 IAM role (for the Lambda to invoke Comprehend and DynamoDB)

Reviewer: * please use "one" instead of "1"
* please use Lambda function instead of Lambda

Author: Done; updated README.md

## Requirements

### Development Environment
**Cloud 9**

> This code for this pattern is executed in AWS Cloud9 environment. The EC2 instance used is m5.large (8 GiB RAM + 2 vCPU).

Reviewer: Why do I need a Cloud9 instance for this? Is it not possible to run this locally as well?

Author: The idea is to use Cloud9 to perform the CDK deployment. To use locally, user may have to use AWS SAM CLI to locally test the AWS CDK applications. 
I've updated the messaging.

Reviewer: Sorry, maybe there is some confusion on my side. Could you reiterate why a Cloud9 environment is necessary and a local executing of `cdk synth && cdk deploy` is not advised? I've managed to deploy it from my local machine just fine. And how does the AWS SAM CLI relate to this?

Author: Apologies, my intention was to keep everything within AWS ecosystem (which includes the CDK synth and deploy). There is nothing blocking users from using local environment. It's just that I used a Cloud9 environment as per the convenience.

Users can also use AWS SAM CLI to deploy CDK apps. 
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-cdk-testing.html

Reviewer: Ok, thanks for the clarifications. I think it would be best to add a line about the possibility to do it locally as well. The Cloud9 instance will incur cost (even if minimal) and users might forget to delete it afterwards. On that note, I don't see a callout to terminate the instance at the end in the README.

Author: Thanks a lot for the observation.
Yes, I've added a comment showcasing capability to execute CDK deployment from local environment and towards the end to remind users to terminate the Cloud9 EC2 instance.

### AWS setup
**Region**

> If you have not yet run `aws configure` and set a default region, you must do so, or you can also run `export AWS_DEFAULT_REGION=<your-region>`. The region used in the demonstration is us-east-1 as Bedrock service is available only in limited regions.

Reviewer: ```suggestion
If you have not yet run `aws configure` and set a default region, you must do so, or you can also run `export AWS_DEFAULT_REGION=<your-region>`. The region used in the demonstration is us-east-1 as Amazon Bedrock is currently not available in every region.
```

**Authorization**

> You must use a role that has sufficient permissions to create Iam roles, as well as cloudformation resources

Reviewer: ```suggestion
You must use a role that has sufficient permissions to create IAM roles, as well as CloudFormation resources
```

#### Python >=3.8

> Make sure you have [python3](https://www.python.org/downloads/) installed at a version >=3.8.x in the CDK environment. The demonstration has used python version 3.10.

Reviewer: ```suggestion
Make sure you have [python3](https://www.python.org/downloads/) installed at a version >=3.8.x in the CDK environment. The demonstration uses python 3.10.
```

#### AWS CDK
Make sure you have the [AWS CDK](https://docs.aws.amazon.com/cdk/v2/guide/getting_started.html#getting_started_install) installed in the Cloud9 environment.

## Setup

### Set up environment and gather packages

```
cd s3-lambda-comprehend-cdk-python
```

Install the required dependencies (aws-cdk-lib and constructs) into your Python environment 
```
pip install -r requirements.txt
```

### Gather and deploy resources with the CDK

> First synthesize, which executes the application, defines which resources will be created, and translates this into a cloudformation template

Reviewer: What do you mean by "executes the application"?

Author: The intention was to help under what the `cdk synth` command does here. 
I've tried to summarize the contents from the AWS documentation here. 

https://docs.aws.amazon.com/cdk/v2/guide/hello_world.html#hello_world_tutorial_synth

> First synthesize, which executes the application, defines which resources will be created, and translates this into a cloudformation template

Reviewer: ```suggestion
First synthesize, which executes the application, defines which resources will be created, and translates this into a CloudFormation template
```

```
cdk synth
```

> Now bootstrap, which provisions the resources you'll use when deploying the application

Reviewer: This is a one-time process and only necessary before deploying your first CDK application

Author: Updated the messaging.

```
cdk bootstrap
```
and deploy with
```
cdk deploy
```

> The deployment will create a S3 bucket, a Lambda and a S3 bucket.

Reviewer: Please use Lambda function and remove the duplicate S3 bucket

## How it works
The S3 bucket acts as a placeholder to upload the text, required for performing Sentiment Analysis. In the demonstration, we use the contents inside `src\user_input_text.txt`. 

> The file is uploaded to S3 bucket, which triggers the Lambda. The Lambda processes the contents within the uploaded file and invokes the Comprehend's BatchDetectSentiment API and the analyzed response from Comprehend service is stored in DynamoDB.

Reviewer: ```suggestion
The file is uploaded to S3 bucket, which triggers the Lambda function. The Lambda function processes the contents within the uploaded file and invokes the Comprehend's `BatchDetectSentiment` API and the analyzed response from Comprehend service is stored in DynamoDB.
```

## Testing
Upon successful deployment of the stack, the Output section would provide the `S3 Bucket` in the CDK environment. 

> Upload the sample file `src\user_input_text.txt` into the `S3 bucket`. The upload action triggers the `Lambda` and the input text is analyzed.

Reviewer: ```suggestion
Upload the sample file `src\user_input_text.txt` into the `S3 bucket`. The upload action triggers the Lambda function and the input text is analyzed.
```

> A response as below is returned from the `Comprehend` service and stored in a `DynamoDB` with file name as the `partition key`. 

Reviewer: ```suggestion
A response as below is returned from Amazon Comprehend and stored in a DynamoDB table with file name as the partition key. 
```

> As per specific use-cases, the responses can be converted into a flat file to be stored in `S3`.

Reviewer: ```suggestion
As per specific use-cases, the responses can be converted into a flat file to be stored in S3.
```
